File: SQL/Expressions/SQLStringConcatExpression.vb
' ___________________________________________________
'
'  © Hi-Integrity Systems 2012. All rights reserved.
'  www.hisystems.com.au - Toby Wicks
' ___________________________________________________
'

Option Strict On
Option Explicit On
Option Infer On

Imports System.Linq

Namespace SQL

    Public Class SQLStringConcatExpression
        Inherits SQLExpression

        Private _left As SQLExpression
        Private _right As SQLExpression

        Public Sub New()


        End Sub

        Public Sub New(ByVal leftString As String, ByVal rightString As String)

            Me.New(New SQLValueExpression(leftString), New SQLValueExpression(rightString))

        End Sub

        Public Sub New(ByVal leftExpression As SQLExpression, ByVal rightExpression As SQLExpression)

            Me.LeftExpression = leftExpression
            Me.RightExpression = rightExpression

        End Sub

        Public Property LeftExpression() As SQLExpression
            Get

                Return _left

            End Get

            Set(ByVal value As SQLExpression)

                If value Is Nothing Then
                    Throw New ArgumentNullException
                End If

                _left = value

            End Set
        End Property

        Public Property RightExpression() As SQLExpression
            Get

                Return _right

            End Get

            Set(ByVal value As SQLExpression)

                If value Is Nothing Then
                    Throw New ArgumentNullException
                End If

                _right = value

            End Set
        End Property

        Friend Overrides Function SQL(ByVal eConnectionType As Database.ConnectionType) As String

            If _left Is Nothing Then
                Throw New ArgumentNullException(Me.GetType.Name & ".LeftExpression")
            ElseIf _right Is Nothing Then
                Throw New ArgumentNullException(Me.GetType.Name & ".RightExpression")
            End If

            Return _left.SQL(eConnectionType) & " + " & _right.SQL(eConnectionType)

        End Function

        Public Shared Function ConcatenateAll(ParamArray sqlExpressions() As SQLExpression) As SQLExpression

            If sqlExpressions.Length < 2 Then
                Throw New ArgumentException("Two or more expressions are required for string concatenation")
            End If

            Dim currentExpression As SQLStringConcatExpression = Nothing

            currentExpression = New SQLStringConcatExpression()
            currentExpression.LeftExpression = sqlExpressions(0)
            currentExpression.RightExpression = sqlExpressions(1)

            For Each sqlExpression In sqlExpressions.Skip(2)

                Dim newExpression As New SQLStringConcatExpression()
                newExpression.LeftExpression = currentExpression
                newExpression.RightExpression = sqlExpression

                currentExpression = newExpression
            Next

            Return currentExpression

        End Function

    End Class

End Namespace

File: SQL/Expressions/SQLStringConcatExpression.cs
// _________________________________________________________________________
//
//  © Hi-Integrity Systems 2012. All rights reserved.
//  www.hisystems.com.au - Toby Wicks
// 
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//	    http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
// _________________________________________________________________________
//

using System.Collections;
using System;
using System.Data;
using System.Linq;
	
namespace DatabaseObjects.SQL
{
	public class SQLStringConcatExpression : SQLExpression
	{
		private SQLExpression left;
		private SQLExpression right;
			
		public SQLStringConcatExpression()
		{
		}
			
		public SQLStringConcatExpression(string leftString, string rightString) 
            : this(new SQLValueExpression(leftString), new SQLValueExpression(rightString))
		{
		}
			
		public SQLStringConcatExpression(SQLExpression leftExpression, SQLExpression rightExpression)
		{
			this.LeftExpression = leftExpression;
			this.RightExpression = rightExpression;
		}
			
		public SQLExpression LeftExpression
		{
			get
			{
				return left;
			}
				
			set
			{
				if (value == null)
					throw new ArgumentNullException();
					
				left = value;
			}
		}
			
		public SQLExpression RightExpression
		{
			get
			{
				return right;
			}
				
			set
			{
				if (value == null)
					throw new ArgumentNullException();
					
				right = value;
			}
		}
			
		internal override string SQL(Serializers.Serializer serializer)
		{
			return serializer.SerializeStringContactExpression(this);
		}

		public static SQLExpression ConcatenateAll(params SQLExpression[] sqlExpressions)
		{
			if (sqlExpressions.Length < 2)
				throw new ArgumentException("Two or more expressions are required for string concatenation");
				
			SQLStringConcatExpression currentExpression = null;
				
			currentExpression = new SQLStringConcatExpression();
			currentExpression.LeftExpression = sqlExpressions[0];
			currentExpression.RightExpression = sqlExpressions[1];
				
			foreach (var sqlExpression in sqlExpressions.Skip(2))
			{
				SQLStringConcatExpression newExpression = new SQLStringConcatExpression();
				newExpression.LeftExpression = currentExpression;
				newExpression.RightExpression = sqlExpression;
					
				currentExpression = newExpression;
			}
				
			return currentExpression;
		}
	}
}
